### PR TITLE
Refurbish deconvolution

### PR DESCRIPTION
**Pull request by BjoernLudwigPTB**
During the refurbishing and debugging of GUM2DFT we improved some parts of PyDynamic as a side-effect. This PR introduces those changes to PyDynamic.

### CODE REVIEW

**Review by mgrub — COMMENTED** [review]
That is a lot of work and the user experience should become much better with all those adjusted docstrings, specific error messages, strong typing and wider test-coverage 👍 

One further note: We now have a "Uncertainty for DFT" and a "uncertainty_for_dft" directory in the examples. Can we merge them?

File: PyDynamic/uncertainty/propagate_DFT.py [hunk 1]
```diff
@@ -680,34 +791,45 @@ def Time2AmpPhase_multi(x, Ux, selector=None):
...
+
+    Raises
+    ------
+    ValueError
+        If dimension of UAP is not even.
     """
-
     N = UAP.shape[0] - 2
-    assert np.mod(N, 2) == 0
+    if np.mod(N, 2) != 0:
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_DFT.py:831]: Resolved by 4c4dce7c.

File: PyDynamic/uncertainty/propagate_DFT.py [hunk 2]
```diff
@@ -793,58 +915,91 @@ def DFT_transferfunction(X, Y, UX, UY):
...
     Returns
     -------
-        X: np.ndarray of shape (2M,)
-            real and imaginary parts of DFT values of deconv result
-        UX: np.ndarray of shape (2M,2M)
-            covariance matrix associated with real and imaginary part of X
+    X : np.ndarray of shape (2M,)
+        real and imaginary parts of DFT values of deconv result
+    UX : np.ndarray of shape (2M,2M)
+        covariance matrix associated with real and imaginary part of X
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_DFT.py]: That's a good hint. Resolved in 1672e3bb. What do you think about the wording?

File: PyDynamic/uncertainty/propagate_DFT.py [hunk 3]
```diff
@@ -793,58 +915,91 @@ def DFT_transferfunction(X, Y, UX, UY):
...
+                f"Each dimension of UH is expected to match the length of H, but "
+                f"H is of length {len(H)} and UH is of shape {UH.shape}."
+            )
+        if UH.shape != UY.shape:
+            raise ValueError(
+                f"The dimensions of UH and UY are expected to match but UH is of shape "
+                f"{UH.shape} and UY is of shape {UY.shape}."
+            )
         N = UH.shape[0] - 2
     else:
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_DFT.py:984]: Good spot again. Resolved by fa457b72.

File: PyDynamic/uncertainty/propagate_DFT.py [hunk 4]
```diff
@@ -793,58 +915,91 @@ def DFT_transferfunction(X, Y, UX, UY):
...
+        if len(UH) != len(H):
+            raise ValueError(
+                f"AmpPhase2DFT: The dimensions of UH and H are expected to match but "
+                f"UH is of length {len(UH)} and H is of length {len(H)}."
+            )
+        if len(UY) != len(Y):
+            raise ValueError(
+                f"AmpPhase2DFT: The dimensions of UH and H are expected to match but "
+                f"UH is of length {len(UH)} and H is of length {len(H)}."
+            )
```
[the remark this answers is not in the document]
Comment (BjoernLudwigPTB) [reply PyDynamic/uncertainty/propagate_DFT.py]: Tackled in 4d1adeb0. What do you think?

**Comment by BjoernLudwigPTB** [comment]
> That is a lot of work and the user experience should become much better with all those adjusted docstrings, specific error messages, strong typing and wider test-coverage 
> 
> One further note: We now have a "Uncertainty for DFT" and a "uncertainty_for_dft" directory in the examples. Can we merge them?

Another good spot! That's resolved by 810bd52f.

**Review by mgrub — APPROVED** [review]
Very nice !

**Pull request merged**